_source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md

```diff
@@ -0,0 +1,362 @@
+
+Because access tokens are generally short-lived (an hour by default), you might [prefer to validate the tokens locally](https://developer.okta.com/docs/guides/validate-access-tokens/go/overview/). This method is slightly less secure because you can't be sure that the access token hasn't been revoked remotely, but on the other hand, you don't have to use your Okta client secret to validate the token locally. Finally, local validation is significantly faster because you don't have to make a request to the Okta authorization server with every API call. You'll have to decide which method is most appropriate for your application.
+
+To validate the access token locally, install the [Okta JWT Python package](https://github.com/adithyasampatoor/okta_jwt):
+
+```bash
+pip install okta_jwt
+pip freeze > requirements.txt
+```
+
+Next, import the package's `validate_token` function and update the `validate` function in your `main.py` file:
+
+```python
+...
+from okta_jwt.jwt import validate_token as validate_locally
+
+...
+
+def validate(token: str = Depends(oauth2_scheme)):
+    try:
+        res = validate_locally(
+            token,
+            config('OKTA_ISSUER'),
+            config('OKTA_AUDIENCE'),
+            config('OKTA_CLIENT_ID')
+        )
+        return bool(res)
+    except Exception:
+        raise HTTPException(status_code=403)
+...
+```
+
+When you call the `/items` endpoint, the API will decode the JWT and validate it locally. The decoded JWTs are cached, so subsequent requests will be faster than the first one.
+
+## Conclusion
+In this post, you've seen how to use FastAPI to build a REST API endpoint that uses an external authorization server to generate and validate access tokens. You've seen some of the key features of FastAPI in action, including dependency injection, the OpenAPI documentation, type hinting, and OAuth implementation.
+
+FastAPI is a great option for building secure and performant backend systems. While there's much more to building a robust production API, including testing, handling POST and PUT endpoints, and connecting to a database for persistence, I hope this tutorial helps you get started.
+
+If you want to learn more about FastAPI, I suggest the following resources:
+- [Fast API from the ground up (video)](https://www.youtube.com/watch?v=3DLwPcrE5mA)
+- [Basic OAuth in FastAPI](https://fastapi.tiangolo.com/tutorial/security/simple-oauth2)
+- [Developing a CRUD API with FastAPI](https://testdriven.io/blog/fastapi-crud/)
+
+To learn more about building an app that can communicate with FastAPI, I suggest using one of these tutorials:
+- [Data Visualization in Angular Using D3.js](https://blog.logrocket.com/data-visualization-angular-d3/)
+- [Building Micro Frontends](https://www.telerik.com/blogs/building-micro-frontends)
+- [Building a Response Timer to Benchmark API Performance](https://dev.to/karllhughes/building-a-response-timer-to-benchmark-api-performance-3k6k)
```

Comment:
Missing a call to action for following `@oktadev` and links to other Python / REST posts.